### PR TITLE
Optimize AsyncClientManager by moving hashing to control plane

### PR DESCRIPTION
**Pull request by DiazAlan**
Commit Message: Optimize AsyncClientManager by moving hashing to control plane
Additional Description: This change removes the hashing of the config away from the data plane on every request to the control plane when the filter callback is created. This is accomplished by creating a wrapper object for the config which precomputes the hash and also serializes it to string ahead of time so that calls to the AsyncClientManager with that wrapper can avoid unnecessary overhead.
Risk Level: Low
Testing: Changed some ext_authz tests to account for changes
Added a benchmark test that observed the following improvement
Benchmark                                         Time             CPU   Iterations
testGetOrCreateAsyncClientWithConfig          27199 us        27182 us            1
testGetOrCreateAsyncClientWithHashConfig      15943 us        15943 us            1

### CODE REVIEW

**Comment by DiazAlan** [comment]
@ggreenway This is a follow up from what we talked about on [28272](https://github.com/envoyproxy/envoy/pull/28272). This PR wraps the precomputed hash and config and sends that to the async client manager. I haven't changed the other filters yet as I want to get feedback on the initial implementation of this. If this looks good and passes, I'll make a follow up PR for the other filters.

**Review by pradeepcrao — COMMENTED** [review]
Thanks so much for working on this Alan!

File: source/common/grpc/async_client_manager_impl.cc [hunk 1]
```diff
@@ -136,12 +137,26 @@ AsyncClientManagerImpl::factoryForGrpcService(const envoy::config::core::v3::Grp
 RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
     const envoy::config::core::v3::GrpcService& config, Stats::Scope& scope,
     bool skip_cluster_check) {
-  RawAsyncClientSharedPtr client = raw_async_client_cache_->getCache(config);
+  const GrpcServiceHashKeyWrapper& key_wrapper = GrpcServiceHashKeyWrapper(config);
```
[the remark this answers is not in the document]
Comment (DiazAlan) [reply source/common/grpc/async_client_manager_impl.cc]: Done. Also changed GrpcServiceHashKeyWrapper to GrpcServiceConfigWithHashKey.

File: source/common/grpc/async_client_manager_impl.h [hunk 2]
```diff
@@ -51,32 +52,34 @@ class AsyncClientManagerImpl : public AsyncClientManager {
...
+    void setCache(const GrpcServiceHashKeyWrapper& key_wrapper,
                   const RawAsyncClientSharedPtr& client);
 
-    RawAsyncClientSharedPtr getCache(const envoy::config::core::v3::GrpcService& config);
+    RawAsyncClientSharedPtr getCache(const GrpcServiceHashKeyWrapper& key_wrapper);
 
   private:
     void evictEntriesAndResetEvictionTimer();
     struct CacheEntry {
       CacheEntry(const envoy::config::core::v3::GrpcService& config,
```
[the remark this answers is not in the document]
Comment (DiazAlan) [reply source/common/grpc/async_client_manager_impl.h]: Done

File: source/common/grpc/async_client_manager_impl.h [hunk 3]
```diff
@@ -51,32 +52,34 @@ class AsyncClientManagerImpl : public AsyncClientManager {
...
+      GrpcServiceHashKeyWrapper key_wrapper_;
       RawAsyncClientSharedPtr client_;
       MonotonicTime accessed_time_;
     };
     using LruList = std::list<CacheEntry>;
-    absl::flat_hash_map<envoy::config::core::v3::GrpcService, LruList::iterator, MessageUtil,
-                        MessageUtil>
-        lru_map_;
     LruList lru_list_;
+    absl::flat_hash_map<GrpcServiceHashKeyWrapper, LruList::iterator> lru_map_wrapper_;
```
[the remark this answers is not in the document]
Comment (DiazAlan) [reply source/common/grpc/async_client_manager_impl.h]: Done

File: envoy/grpc/async_client_manager.h [hunk 4]
```diff
@@ -54,6 +75,10 @@ class AsyncClientManager {
   getOrCreateRawAsyncClient(const envoy::config::core::v3::GrpcService& grpc_service,
                             Stats::Scope& scope, bool skip_cluster_check) PURE;
 
+  virtual RawAsyncClientSharedPtr
+  getOrCreateRawAsyncClientWithWrapper(const GrpcServiceHashKeyWrapper& grpc_service,
```
[the remark this answers is not in the document]
Comment (DiazAlan) [reply envoy/grpc/async_client_manager.h]: Done

File: envoy/grpc/async_client_manager.h [hunk 5]
```diff
@@ -32,6 +32,27 @@ class AsyncClientFactory {
...
+  GrpcServiceHashKeyWrapper(const envoy::config::core::v3::GrpcService& config)
+      : config_(config), pre_computed_hash_(Envoy::MessageUtil::hash(config)){};
+
+  template <typename H> friend H AbslHashValue(H h, const GrpcServiceHashKeyWrapper& wrapper) {
+    return H::combine(std::move(h), wrapper.pre_computed_hash_);
+  }
+
+  friend bool operator==(const GrpcServiceHashKeyWrapper& lhs,
+                         const GrpcServiceHashKeyWrapper& rhs) {
+    return Protobuf::util::MessageDifferencer::Equivalent(lhs.config_, rhs.config_);
```
[the remark this answers is not in the document]
Comment (DiazAlan) [reply envoy/grpc/async_client_manager.h]: Done

File: envoy/grpc/async_client_manager.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (DiazAlan) [reply envoy/grpc/async_client_manager.h]: I also just added a comment to the new getOrCreateRawAsyncClient that better describes it's use.

File: envoy/grpc/async_client_manager.h [hunk 6]
```diff
@@ -32,13 +32,38 @@ class AsyncClientFactory {
 
 using AsyncClientFactoryPtr = std::unique_ptr<AsyncClientFactory>;
 
+class GrpcServiceConfigWithHashKey {
+public:
+  GrpcServiceConfigWithHashKey(const envoy::config::core::v3::GrpcService& config)
```
Comment (pradeepcrao) [inline envoy/grpc/async_client_manager.h:37]: nit: Please mark constructor as `explicit`.

**Review by pradeepcrao — COMMENTED** [review]
Looks great Alan! 

Could you please add a unit test for `GrpcServiceConfigWithHashKey`? Also, I feel it might be easier (and better) to write a performance benchmark that compares `getOrCreateRawAsyncClientWithHashKey` with `getOrCreateRawAsyncClient` than try to load test this.

**Comment by phlax** [comment]
@DiazAlan i think this may be waiting for further iteration

/wait

**Review by ggreenway — DISMISSED** [review]
This looks great! Thanks!

**Review by ggreenway — APPROVED** [review]

**Pull request merged**